google/cloud/spanner/admin/integration_tests/instance_admin_rest_integration_test.cc

```diff
@@ -0,0 +1,430 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/spanner/admin/instance_admin_client.h"
+#include "google/cloud/spanner/admin/internal/database_admin_option_defaults.h"
+#include "google/cloud/spanner/admin/internal/database_admin_rest_connection_impl.h"
+#include "google/cloud/spanner/admin/internal/database_admin_rest_stub_factory.h"
+#include "google/cloud/spanner/admin/internal/instance_admin_option_defaults.h"
+#include "google/cloud/spanner/admin/internal/instance_admin_rest_connection_impl.h"
+#include "google/cloud/spanner/admin/internal/instance_admin_rest_stub_factory.h"
+#include "google/cloud/spanner/create_instance_request_builder.h"
+#include "google/cloud/spanner/testing/cleanup_stale_instances.h"
+#include "google/cloud/spanner/testing/pick_instance_config.h"
+#include "google/cloud/spanner/testing/random_instance_name.h"
+#include "google/cloud/spanner/update_instance_request_builder.h"
+#include "google/cloud/common_options.h"
+#include "google/cloud/internal/getenv.h"
+#include "google/cloud/internal/random.h"
+#include "google/cloud/internal/rest_background_threads_impl.h"
+#include "google/cloud/testing_util/integration_test.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include "absl/strings/match.h"
+#include <gmock/gmock.h>
+#include <algorithm>
+#include <string>
+#include <vector>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using ::google::cloud::testing_util::IsOk;
+using ::google::cloud::testing_util::StatusIs;
+using ::testing::AnyOf;
+using ::testing::EndsWith;
+using ::testing::Eq;
+using ::testing::HasSubstr;
+using ::testing::IsEmpty;
+using ::testing::Not;
+using ::testing::UnorderedElementsAre;
+
+std::string const& ProjectId() {
+  static std::string project_id =
+      internal::GetEnv("GOOGLE_CLOUD_PROJECT").value_or("");
+  return project_id;
+}
+
+std::string const& InstanceId() {
+  static std::string instance_id =
+      internal::GetEnv("GOOGLE_CLOUD_CPP_SPANNER_TEST_INSTANCE_ID")
+          .value_or("");
+  return instance_id;
+}
+
+bool RunSlowInstanceTests() {
+  static bool run_slow_instance_tests = absl::StrContains(
+      internal::GetEnv("GOOGLE_CLOUD_CPP_SPANNER_SLOW_INTEGRATION_TESTS")
+          .value_or(""),
+      "instance");
+  return run_slow_instance_tests;
+}
+
+bool Emulator() {
+  static bool emulator =
+      internal::GetEnv("SPANNER_EMULATOR_REST_HOST").has_value();
+  return emulator;
+}
+
+// We need to override stub creation to use an alternate emulator endpoint for
+// the InstanceAdmin.
+std::shared_ptr<google::cloud::spanner_admin::InstanceAdminConnection>
+MakeInstanceAdminConnectionRestEmulator() {
```

Comment:
Why shouldn't `MakeInstanceAdminConnectionRest()` deal with all this?  (And similarly for `DatabaseAdmin`.)

Reply:
I think it is the `emulator_endpoint_end_var`.

https://github.com/googleapis/google-cloud-cpp/blob/10cfef588c819438b85c5a086fd9f44d8f087288/generator/generator_config.textproto#L1452

We could add a generator flag that supersedes that env var for REST... or split up the REST and gRPC into separate `service`s in the `generator_config.textproto`.

Reply:
As Darren suggests, the issue with with needing an additional env var for the rest emulator port. As only spanner currently needs this workaround, I'm not planning on changing the generator at this time.